### ISSUE TITLE
Training going very slow

### ISSUE THREAD

**Issue by Vijetha1** [fchollet/keras#1653]
I am trying to train a model built with Keras's Sequential class. The training is very slow with 8000 secs per epoch, i.e 9.25 days for 100 epochs which I can't afford to wait. 

Details about the model and data are as follows. 
1. The model has 2 channels and each one has 4 CNN layers, 2 max-pooling layers and 3 dense layers.
2. Both the channels are independent in the initial layers and are combined only before the final dense layer. 
3. The input size is (150000, 3, 224, 224) with 32 samples per gradient update. I have 2 different data for the 2 channels and therefore have a total of 2*150,000 images sent to the network in pairs, one for each channel. 
4. I am using all `relu` activations but a `softmax` one in the last but one layer; optimizer is `rmsprop` and loss is `categorical_crossentropy`

Details about the setup is as follows.
1. Ubuntu - 14.04.3
2. GPU - Nvidia Tesla K40c(GK110BGL)/33Mhz/12GB
3. Cuda - 7.5
4. Python - 2.7.6
5. Keras backend - Theano

The code-piece where I train the model is given here. 

```
h5fGlobal = h5py.File( 'global_data.h5', 'r')
h5fLocal = h5py.File( 'local_data.h5', 'r')
h5fLabel = h5py.File( 'label_data.h5', 'r')
label_data = h5fLabel['label_data'][:]
h5fLabel.close()
label_data = np_utils.to_categorical(label_data,2)
model.fit(X=[h5fGlobal['global_data'], h5fLocal['local_data']],
                  y=label_data, batch_size=32, nb_epoch=1,
                  callbacks=[], validation_split=0.0, validation_data=None,  shuffle="batch",show_accuracy=False,
                  class_weight=None, sample_weight=None)
```

I put the entire data into 8 global and 8 local files(the size of each file is 22GB) and run the `fit` function on each of these. The program loops through them for 100 epochs.  

My ~.theanorc file is given below in case you wish to look at it. 

```
[global]
floatX = float32
device=gpu0
exception_verbosity=high
optimizer_including=cudnn
mode=FAST_RUN
openmp=True

[blas]
ldflags = -L/usr/lib -lblas -llapack

[nvcc]
nvcc.fastmath=True
allow_gc=True

[cuda]
root = /usr/local/cuda-7.5/

[lib]
cnmem=0.85

[gcc]
cxxflags=-I /usr/local/cuda/include -L /usr/local/cuda/lib64

[scan]
allow_gc=True
allow_output_prealloc=True

```

Could you give me a clue why my training is so slow?

**Comment by fchollet** [comment]
> Could you give me a clue why my training is so slow?

Almost certainly because you don't have cuDNN installed (or otherwise it's not running).

**Comment by Vijetha1** [comment]
I installed it quite sometime back. I can reinstall it though to be on a safer side (or verify if it is installed). But how do I check if it is running when I run my program? Can you help?

**Comment by Vijetha1** [comment]
We just found out a way to check if cuDNN is running from 
http://www.chioka.in/why-is-keras-running-so-slow/

We will let you know how it goes. Thank you!

**Comment by Vijetha1** [comment]
I verified the installation of cuda and cuDNN and everything looks fine. 

I tried on two different systems and it is taking around 7000-8000 seconds per epoch on both the systems. Can someone explain me how long would it generally take for network this deep(it has 2 columns) and data of size (150,000*2, 3, 224, 224 )(multiplied by 2 because I have 2 channels)? When I analyse the time, around 50%(a little more than 50%) is utilized in loading the files and a little less than 50% is utilized in training. (I just have 2 parts, load and train, no preprocessing as I preprocessed when I made the hdf5 files).

@fchollet Can you help me with this? I am precisely looking for a benchmark to compare my time-per-epoch with.

Thank you.

**Comment by easyas314159** [comment]
@fchollet @Vijetha1 I just ran into a similar problem and after running some tests I am pretty sure the issue is with how Keras loads samples from HDF5 datasets.

All of the indices passed to the HDF objects `__getitem__` method use fancy indexing which is significantly slower than using a slice. The [fancy indexing section](http://docs.h5py.org/en/latest/high/dataset.html#fancy-indexing) of the h5py docs allude to this but I didn't appreciate how much of a performance bottleneck it was until I ran some benchmarks.

On my system using a relatively simple random dataset of shape `(10000, 100)`, dtype float32, and no compression:
- Load 1000 rows using a slice (e.g. `X[0:1000]`): 4.094ms average
- Load 1000 rows using fancy indexing (e.g. `X[np.arange(1000)]`): 254.588ms average

A slow down by a factor of ~62x is pretty significant. Doubling the number of rows loaded to 2000 increases the performance discrepancy even more to a whopping ~118x slow down (870.254ms vs. 7.378ms). At 4000 rows it's a staggering ~176x slow down (2525.599ms vs. 14.369ms). Even on relatively small requests for 10 rows the slowdown is still ~9x.

I would argue that this is a pretty significant performance issue that should probably be fixed. If someone points me at where the fancy index is passed to the train/test data I can put together a pull request.
